django/core/mail/message.py

```diff
@@ -308,10 +308,32 @@ def attach(self, filename=None, content=None, mimetype=None):
             self.attachments.append((filename, content, mimetype))
 
     def attach_file(self, path, mimetype=None):
-        """Attaches a file from the filesystem."""
+        """
+        Attaches a file from the filesystem.
+
+        The mimetype will be set to the DEFAULT_ATTACHMENT_MIME_TYPE if it is
+        not specified and cannot be guessed or (PY3 only) if it suggests
+        text/* for a binary file.
+        """
         filename = os.path.basename(path)
-        with open(path, 'rb') as f:
+        if not mimetype:
+            mimetype, _ = mimetypes.guess_type(filename)
+            if not mimetype:
+                mimetype = DEFAULT_ATTACHMENT_MIME_TYPE
+        basetype, subtype = mimetype.split('/', 1)
+        read_mode = 'r' if basetype == 'text' else 'rb'
+        f = open(path, read_mode)
```

Comment:
I replaced context manager version with open/close approach, to avoid opening the same file twice, but with different mode. But maybe inner function could be helpful here?